src/TodosProvider.tsx

```diff
@@ -0,0 +1,58 @@
+import { useMemo, useState } from 'react';
+import { TodosContext } from './TodosContext';
+import { Error } from './types/Error';
+import { Todo } from './types/Todo';
+import { apiActions } from './api/apiActions';
+
+type Props = {
+  children: React.ReactNode;
+};
+
+export const TodosProvider: React.FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [error, setError] = useState(Error.None);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [deletingTodosId, setDeletingTodosId] = useState<number[]>([]);
+
+  function deleteTodo(id: number) {
+    return apiActions.delete(id)
+      .then(() => {
+        setTodos(currentTodos => currentTodos
+          .filter(item => item.id !== id));
+      })
+      .catch(() => {
+        setTodos(todos);
+        setError(Error.Delete);
+      });
+  }
+
+  function addTodo({ title, completed, userId }: Omit<Todo, 'id'>) {
+    return apiActions.add({ title, completed, userId })
+      .then(newTodo => {
+        setTodos(currentTodos => [...currentTodos, newTodo]);
+      })
+      .catch((e) => {
+        setError(Error.Add);
+        throw e;
+      });
+  }
+
+  const state = useMemo(() => ({
+    todos,
+    tempTodo,
+    setTodos,
+    error,
+    deletingTodosId,
+    setError,
+    setTempTodo,
+    deleteTodo,
+    addTodo,
+    setDeletingTodosId,
+  }), [todos, error, tempTodo, deletingTodosId]);
```

Comment:
Don't overuse Context, it requires more resources than simple state lifting. 
Please read the [doc](https://legacy.reactjs.org/docs/context.html#:~:text=Context%20is%20primarily%20used%20when,a%20simpler%20solution%20than%20context.) 
You can also google other resources about it to figure out the theme. And feel free to ask in the chat if you have questions

Reply:
In simple words, use Context when you need to pass props through components without using them, or for parameters that are needed in all components(for ex. language)